scanpipe/pipes/scancode.py

```diff
@@ -482,40 +485,210 @@ def process_package_data(project):
                     project=project,
-                    dependency_data=dep,
-                    for_package=package,
+                    dependency_data=dep_from_resolved,
+                    for_package=resolved_to_package,
                     datafile_resource=resource,
-                    datasource_id=pd.datasource_id,
+                    datasource_id=datasource_id,
                 )
 
+        pipes.update_or_create_dependency(
+            project=project,
+            dependency_data=dep,
+            for_package=package,
+            resolved_to_package=resolved_to_package,
+            datafile_resource=resource,
+            datasource_id=datasource_id,
+        )
+
+
+def match_and_resolve_dependencies(project):
+    """
+    From a project with both direct dependency relationships (contains
+    only the parent package and the requirement) and indirect dependency
+    relationships like in lockfiles (this contains the resolved package
+    and the requirement), match and update dependencies to contain the
+    full dependency graph.
+    """
+    for dependency in project.discovereddependencies.all():
+        if dependency.resolved_to_package:
+            continue
+
+        purl_data = DiscoveredDependency.extract_purl_data(
+            dependency_data={"purl": dependency.purl},
+            ignore_nulls=True,
+        )
+        extracted_requirement = dependency.extracted_requirement
+        if not extracted_requirement:
+            extracted_requirement = ""
+
+        matched_dependencies = DiscoveredDependency.objects.filter(
+            project=project,
+            extracted_requirement=dependency.extracted_requirement,
+            **purl_data,
+        )
+
+        other_dependencies = [
+            dep for dep in matched_dependencies if dep.purl != dependency.purl
+        ]
+        if not other_dependencies:
+            # We also have cases where multiple dependency requirements have one
+            # resolved package and the extracted requirements field is combined
+            matched_dependencies = DiscoveredDependency.objects.filter(
+                project=project,
+                **purl_data,
+            )
+            other_dependencies = [
+                dep
+                for dep in matched_dependencies
+                if (
+                    dep.purl != dependency.purl
+                    and dependency.extracted_requirement in dep.extracted_requirement
+                )
+            ]
+
+            # This should be done only in the case of lockfiles where only one version
+            # of a package is present for an environment
+            if not other_dependencies:
+                other_dependencies = [
+                    dep
+                    for dep in matched_dependencies
+                    if (
+                        dep.base_purl == dependency.base_purl
+                        and dep.resolved_to_package
+                    )
+                ]
```

Comment:
The var naming is not great and makes the code hard to read: eg: `dep.purl != dependency.purl`

Reply:
Renaming `dep` to `matched_dependency`, thanks!